### ISSUE TITLE
Store Isolated Node in Blacklist

### ISSUE THREAD

**Issue by alschmut** [MaibornWolff/codecharta#268]
# Feature Request

The reason to go for this feature was actually the bug of hiding/showing nodes within the TreeView and  the blacklist. Instead of doing a workaround for the bug I decided to create this feature request.

### Description

When isolating a node we had a conflict between:
- changing the current visibility state of all nodes and 
- setting up a completely new state (like we do since using the blacklist)

So to be able to also isolate a node, we could store that node inside the blacklist as well. It is not possible to isolate more than one node (as that doesn't make sense). When rendering the new map, it now can read from the blacklist if and which node is supposed to be isolated.

Isolating a node does following:
1. keep the excluded nodes excluded
2. keep the hidden nodes hidden
3. only show the isolated nodes with its visible children

Good side-effect:
When the user wants to download that map for a later point, it also saves the isolated node and shows it just as the user left it.

### Question
Do you think **blacklist** is still the correct word and if not do you have a better idea?

**Comment by felixherbst** [comment]
I like the solution, but i think we definitely need a new name, as blacklist suggests that none of that nodes will be visible anymore and isolate shouldn't show anything but that "blacklisted" node, correct me if I'm wrong. Something like "processed nodes" would be more universal, but if we change that name, we lose our compatibility... Therefore wouldn't it be better to just add a new attribute to cc.json instead of sharing the blacklist?

**Comment by alschmut** [comment]
> isolate shouldn't show anything but that "blacklisted" node

Can you explain that? ;)

As the blacklist is kind of fixed in the cc.json anotger lost to hold the other list items is not too bad. I am still thinking about another name. What do you think about **node behavio(u)r**?

**Comment by felixherbst** [comment]
I meant that a blacklist should hide everything that is listed inside. But now, we have that additional "isolated" node inside, which is the exact opposite, so it is visible while everything else is hidden.

**Comment by Richargh** [comment]
I don't really understand the idea behind storing the isolated node in the cc.json. Why would a user do this? I always use isolate feature to focus on something and then go back to the big view. I don't see why it makes sense to store it permanently.

**Comment by Richargh** [comment]
So this is apparently being merged. I would really like to know the usecase before this goes on master because I just don't see the purpose.

**Comment by Richargh** [comment]
Side-note, a bug usually does not result in a new feature and in a change to the cc.json.

**Comment by alschmut** [comment]
Yes, the feature request did not come from a user, but that doesn't mean there is no use-case for it. For the same reason you would like to exclude or hide nodes and save the cc.json file, you might want to isolate only a part of the map, which is on focus at the moment. When reloading the file you see exactly what you saw before. I think as a user its also quite handy to see the isolate item somewhere and be able to delete it again.

> So this is apparently being merged
Nothing was merged?!

Of course we can implement this in another way! I just thought it perfectly fits to the existing kind of way we store the node-behaviour.

-> The actual problem is, that the blacklist items (exclude or hide) do not update the node-visibility itself, but only store the items in `settings.blacklist` and when reloading the map it gets applied. But that is in conflict with the Isolate functionality, as this gets applied directly to the map and is not stored anywhere. So its not reproducible when the map was reloaded (onSettingsChanged). 

So another idea would be to store the isolated node somewhere else in the application (without the user seeing or storing it).

**Comment by Richargh** [comment]
Most important: it's awesome if you suggest a feature. It was never my intent to critique you. 

I question if a user would ever use the feature. In fact it's probably a good idea not to have it in the cc.json. Then he can just reload the map if he ever get's confused that he only sees a part of it. Granted that is a very weak reason :) 

The reason I don't see it in the file is two-fold: a) no conversation I have had with users suggested they need this feature. b) the technical discussion about naming: "does it belong in the blacklist". Since the feature isolates something it does not seem to belong in the blacklist. So it probably needs a new element in the cc.json. Which brings us back to, does someone need it? 

------

--> the actual problem. I'm going to check it out right now.

**Comment by alschmut** [comment]
@Richargh Thanks for your detailed explanation ;) I quite like your suggestion! and would like create a new issue for that. So we can close this one here.

**Comment by Richargh** [comment]
Sure, please create an issue.
